### PR TITLE
enhancement(database)!: add db identifier shortening and migration

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- Enables the name shortening algorithm in the database
- adds an internal migration to ensure all identifiers are renamed as gracefully as possible so no data will be lost (note that foreign key indexes are currently not renamed properly, and in the case of sqlite, all indexes must be dropped and recreated because they do not support renaming indexes)

#### TODO

These will likely come in another PR before v5 release.

- [ ] find a way to rename foreign key indexes so that they don't have to be recreated by the db schema sync
- [ ] add more tests, particularly testing snake case names in relations

### Why is it needed?

To finally close https://github.com/strapi/strapi/issues/13117 🥳

### How to test it?

Simple version: starting with a Strapi 4 project, switch to this branch and start Strapi. The migration should run, and any of your database identifiers (table names, column names, indexes) that were longer than 55 characters will be renamed to fit into that length. [Additionally, identifier parts such as "components" will have been shortened to "cmp", "order_inv_fk" to "oifk", etc.
](https://github.com/strapi/strapi/pull/19783)

Detailed version:

What is currently even possible to test will depend on the database being used. By default, postgres has a limit of 63 characters and mysql has a limit of 65, sqlite is unlimited.

- Start with a Strapi 4 project (from `develop`)
- add some content types with names that are at least 56 characters long
- add some attribute names (of all types, including relations, components, and dynamic zones) that are at least 56 characters long (postgres+mysql might require being in a content type name that is only a couple characters for that to work, it will take some trial and error to find the right balance for what strapi 4 will accept)
- in those, include some names that include numbers and/or other things that will test snakeCasing
- add a full range of content to each of those
- [sqlite only] create content types and attributes with names that are > 63 characters long
- at this point, you might want to do a dump of the database so that you can restore to it and test again later
- Switch to this branch + build
- Start up the Strapi project
- no data should have been lost
- now create content types and attributes of all types with names longer than 65 characters (the previous hard limit within the dbs)
- they should all work!

You should also see the following changes in your database:
- every table name longer than 55 characters should be shortened to 55
- every column name longer than 55 characters should be shortened to 55
- every index name longer than 55 characters should be shortened to 55
- prefixes and suffixes should also be using shortened forms, eg "oifk" instead of "order_inv_fk"
- absolutely no data loss should happen

If you notice any identifiers in the database longer than 55 characters, or any data loss, please report details here.

Additionally, tests have already been provided in previous PRs for all the naming functions and `createMetadata` itself, covering every case we could think of comparing the full length to the expected shortened length, but since there are potentially edge cases that were missed, if you notice any please let me know.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (Bassel17) [inline packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:27]: you want to remove this I believe

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (Bassel17) [inline packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:32]: is it possible for this to happen ? and if does the users are supposed to add the db config ?

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (Bassel17) [inline packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:75]: we don't intend to right?

**Review by Bassel17 — CHANGES_REQUESTED** [review]
looks good, just have some comments on the migration

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (innerdvations) [reply packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:75]: nope

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (innerdvations) [reply packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:32]: Nope, I just removed the option to change the max length since it was dangerous for no benefit and this got left in. Nice catch.

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (innerdvations) [reply packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:27]: Oops, thanks!

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (Convly) [inline packages/core/core/src/utils/transform-content-types-to-models.ts:17]: If we're not using the options, why are we receiving them here? What prevents us from removing the options + the eslint disable rule?

Comment (Convly) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:17]: Same for the other ones below. Although if I remember correctly, those constants functions will be replaced by new ones in the other identifiers PR, right?

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (Convly) [inline packages/core/core/src/utils/transform-content-types-to-models.ts:30]: Same as above, this is handled by the other PR, correct?

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (Convly) [inline packages/core/core/src/utils/transform-content-types-to-models.ts:297]: So the model still has the long name, but the db metadata has a short one? Will the original model be modified or not?

File: packages/core/database/src/metadata/__tests__/resources/models.ts
Comment (Convly) [inline packages/core/database/src/metadata/__tests__/resources/models.ts:225]: What are we doing here? Creating a model with additional attributes?

File: packages/core/database/src/metadata/index.ts
Comment (Convly) [inline packages/core/database/src/metadata/index.ts:28]: Do we have default values for the metadata options? Or should this always be defined manually?

File: packages/core/database/src/metadata/metadata.ts
Comment (Convly) [inline packages/core/database/src/metadata/metadata.ts:46]: Do we still have the issue where the modifying the indexes in the metadata will also update the indexes in the models because of how we're dealing with the referenced array? Or was it coming from somewhere else?

File: packages/core/database/src/metadata/metadata.ts
Comment (Convly) [inline packages/core/database/src/metadata/metadata.ts:101]: Nice catch

File: packages/core/database/src/metadata/relations.ts
Comment (Convly) [inline packages/core/database/src/metadata/relations.ts:88]: It feels a bit weird having all those options, especially having meta, metadata and metadata options next to each others. but that's an issue for another day

Comment (Convly) [reply packages/core/database/src/metadata/relations.ts:88]: We could(/probably should) limit the number of arguments we can pass to a function in our eslint conf

File: packages/core/database/src/metadata/relations.ts
Comment (Convly) [inline packages/core/database/src/metadata/relations.ts:389]: Why do we keep this here if we're not using it?

File: packages/core/database/src/metadata/relations.ts
Comment (Convly) [inline packages/core/database/src/metadata/relations.ts:411]: Same

File: packages/core/database/src/schema/index.ts
Comment (Convly) [inline packages/core/database/src/schema/index.ts:38]: Could we add a bit more detail about which schema we're creating? Maybe adding the uid somewhere in the log?

File: packages/core/database/src/utils/identifiers/shortener.ts
Comment (Convly) [inline packages/core/database/src/utils/identifiers/shortener.ts:29]: Could we split this as
```ts
export type CompressibleNameToken = {
  compressible: true;
  name: string;
  allocatedLength?: number;
}

export type IncompressibleNameToken = { 
  compressible: false;
  name: string;
  allocatedLength?: number;
  shortName?: string;
}

export type NameToken = CompressibleNameToken | IncompressibleNameToken
```

So that we can discriminate on the compressible attribute and avoid making mistakes by using the shortName value for compressible tokens?

File: packages/core/database/src/index.ts
Comment (Convly) [inline packages/core/database/src/index.ts:61]: Did we switch from 64 to 55?

**Review by Convly — COMMENTED** [review]
I haven't looked at the migration yet, but here's the first part of the review

File: packages/core/database/src/index.ts
Comment (innerdvations) [reply packages/core/database/src/index.ts:61]: 63 is the hard maximum for postgres, we went with 55 to reserve a buffer in case we need to add table prefixes, for example (like for environments, multitenancy, etc)

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (innerdvations) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:17]: Yeah, originally this was going to be needed to determine the short versions. It will be removed once the other PR is merged into this one, so for now I'll leave this to avoid adding merge conflicts for something that will be replaced.

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (innerdvations) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:30]: Same answer as above

File: packages/core/database/src/metadata/__tests__/resources/models.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/__tests__/resources/models.ts:225]: @christiancp100 added this so maybe he can clarify, but yeah, this is basically a test utility to simplify building models to be tested

File: packages/core/database/src/metadata/index.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/index.ts:28]: We made it required, because accidentally calling it without options in even one place could cause data loss. In a later refactoring I want to remove MetadataOptions entirely, because it should be possible to do something like instantiate an identifiers object in database with a fixed max length and export that to be used everywhere, but I couldn't come up with a clean pattern for it and ran out of time. But that cleanup can be done without breaking changes.

File: packages/core/database/src/metadata/metadata.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/metadata.ts:46]: The issue was either caused by something I did, or (my suspicion) is actually still an issue but doesn't affect the migration

File: packages/core/database/src/metadata/metadata.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/metadata.ts:101]: So difficult to debug and find a solution for, and yet so simple once I found it 😆

File: packages/core/database/src/metadata/relations.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/relations.ts:88]: Mentioned in another comment; I agree, and my goal is to eliminate the options entirely, but once it was done, working, and tested, I didn't want to do another major refactoring last minute.

File: packages/core/database/src/metadata/relations.ts
Comment (innerdvations) [reply packages/core/database/src/metadata/relations.ts:389]: It was left as a reminder that they're there and must be used for names, but I can remove them and maybe add a comment where it makes sense.

Comment (innerdvations) [reply packages/core/database/src/metadata/relations.ts:389]: Although it might be worth leaving for consistency, what do you think? There are only these two out of eight methods that aren't currently using them

File: packages/core/database/src/schema/index.ts
Comment (innerdvations) [reply packages/core/database/src/schema/index.ts:38]: It's not a specific schema, it's the entire database schema that is all done at once, but I'll clarify that

File: packages/core/database/src/utils/identifiers/shortener.ts
Comment (innerdvations) [reply packages/core/database/src/utils/identifiers/shortener.ts:29]: Gladly, thanks!

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (innerdvations) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:297]: Oooh, nice catch, this comment was a lie / left over from when metadata did shorten the table names. But that has changed and now this has to be shortened. I'm fixing this and will try to come up with some better tests too.

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (innerdvations) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:17]: In fact, I'll go ahead and merge that in to this one now. I think it's been separate long enough to be reviewed but now it's getting confusing.

File: packages/core/core/src/utils/transform-content-types-to-models.ts
Comment (innerdvations) [reply packages/core/core/src/utils/transform-content-types-to-models.ts:317]: Update: no, the code was correct, the comment was still misleading. it's handled in loadModels() not createMetadata(), so I will specify that so it's clear

File: packages/core/database/src/metadata/relations.ts
Comment (Convly) [reply packages/core/database/src/metadata/relations.ts:389]: I'm a bit annoyed to see arguments we don't need being declared, but it's not the biggest deal

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (Convly) [inline packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:73]: Why not reusing `IdentifiersDiff['indexes']`?

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (Convly) [inline packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:162]:
```suggestion
  shortArr.forEach(([, shortObj], index) => {
```

SInce you don't need to keys, why aren't you doing an Object.values above?

Comment (innerdvations) [reply packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts:162]: Haven't you seen I love keeping around variables for posterity? 😆

File: packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts
Comment (innerdvations) [reply packages/core/database/src/migrations/internal-migrations/5.0.0-convert-identifiers-long-than-max-length.ts]: It was an array, but I cleaned up that it was a duplicate

**Review by Convly — APPROVED** [review]
Need some QA, but the code LGTM

**Review by Bassel17 — APPROVED** [review]
Looks good !

**Comment by derrickmehaffy** [comment]
Got this warning on booting up using MariaDB: 11.2.3-MariaDB

```
$ strapi develop
⠋ Loading Strapi⠋ Building build context
⠙ Loading Strapi[INFO] Including the following ENV variables as part of the JS bundle:
    - ADMIN_PATH
    - STRAPI_ADMIN_BACKEND_URL
    - STRAPI_TELEMETRY_DISABLED
✔ Building build context (15ms)
✔ Creating admin (394ms)
⠧ Loading Strapi{
  event: 'migrating',
  name: '5.0.0-rename-identifiers-longer-than-max-length'
}
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
⠇ Loading StrapiTransaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
{
  event: 'migrated',
  name: '5.0.0-rename-identifiers-longer-than-max-length',
  durationSeconds: 0.054
}
✔ Loading Strapi (8347ms)
✔ Generating types (414ms)
```

Looks like it's referencing:

https://github.com/knex/knex/issues/1359
https://github.com/knex/knex/issues/805

File: packages/core/database/src/schema/schema.ts
Comment (alexandrebodin) [inline packages/core/database/src/schema/schema.ts:37]: To avoid exposing  & passing arround the metadata options could you store them in the metadata and expose a meta.getOptions() fn instead to keep a bit of encapsulation

Comment (alexandrebodin) [reply packages/core/database/src/schema/schema.ts:37]: Or it's a matter of using another type :)

File: packages/core/database/src/utils/identifiers/shortener.ts
Comment (alexandrebodin) [inline packages/core/database/src/utils/identifiers/shortener.ts:76]: A few thoughts

- We couble be able to remove that map in a later release as it is only necessary to run the migration 🤔 
- Should we clear the map at the end of the db init phase

**Review by alexandrebodin — COMMENTED** [review]
Just did a very quick read. pretty nice 👍 

Note:  We definitely want to create an identifierProvider that we can init with options instead of passing the options in each function call :)

File: packages/core/database/src/utils/identifiers/shortener.ts
Comment (innerdvations) [reply packages/core/database/src/utils/identifiers/shortener.ts:76]: Yeah, we can remove big parts of this in a later release as long as we require migrating to 5.0.0 (or whatever) first.

I think leaving the map could be potentially useful, it doesn't use much memory. But I also don't mind removing it, the use cases for it are pretty limited.

**Comment by innerdvations** [comment]
Just did some last minute QA and there are some more indexes that aren't being renamed properly, at least on sqlite. If it's -only- on sqlite that's fine and we can ignore it, because they're not actually renamed anyway, but if it's mysql and postgres I would like to fix that to avoid having to recreate indexes unneccessarily. I'll look into it on Monday

File: packages/core/database/src/utils/identifiers/shortener.ts
Comment (alexandrebodin) [reply packages/core/database/src/utils/identifiers/shortener.ts:76]: Yeah if we see a need to remove it we can but not worth planning for it :)

File: packages/core/database/src/metadata/metadata.ts
Comment (innerdvations) [inline packages/core/database/src/metadata/metadata.ts:20]: @alexandrebodin I've reduced the entire public footprint down to this. And like I've written in the comment, in a later refactor this should be completely encapsulated with its own identifiers object, but that would require refactoring every method called by metadata to be part of a metadata instance so they could access the local one, otherwise we're back at the problem of needing to pass an additional argument to every method here.

But at least like this, the public interface is as minimal and correct as it can be and it won't be a breaking change if we want to refactor later.

**Review by Convly — APPROVED** [review]

**Comment by gani-lw** [comment]
@innerdvations @Convly  Why can't this feature be optional? our database is connected with external BI tools like Metabase and now if we upgrade from v4 to v5 we will have to update all the queries, which is quite challenging. 

There should be a config to make it optional. By default it can be enabled but there should be optional to ignore the shortening

**Comment by innerdvations** [comment]
> There should be a config to make it optional. By default it can be enabled but there should be optional to ignore the shortening

It might be possible. Could you create a feature request for it? This is a very old PR, it won't get any traction here :)

**Pull request merged**
